pkg/tests/tasks/traffic/traffic_tcp_shifting_test.go

```diff
@@ -15,100 +15,201 @@
 package traffic
 
 import (
-	_ "embed"
 	"fmt"
 	"strings"
 	"testing"
-	"time"
 
-	"github.com/maistra/maistra-test-tool/pkg/examples"
+	"github.com/maistra/maistra-test-tool/pkg/app"
 	"github.com/maistra/maistra-test-tool/pkg/util"
-	"github.com/maistra/maistra-test-tool/pkg/util/log"
+	"github.com/maistra/maistra-test-tool/pkg/util/hack"
+	"github.com/maistra/maistra-test-tool/pkg/util/oc"
+	"github.com/maistra/maistra-test-tool/pkg/util/pod"
+	"github.com/maistra/maistra-test-tool/pkg/util/retry"
 	"github.com/maistra/maistra-test-tool/pkg/util/test"
 )
 
-var (
-	// OSSM need custom changes in VirtualService tcp-echo
-	//go:embed yaml/tcp-echo-all-v1.yaml
-	echoAllv1Yaml string
-
-	//go:embed yaml/tcp-echo-20-v2.yaml
-	echo20v2Yaml string
-)
-
-func cleanupTCPShifting() {
-	log.Log.Info("Cleanup")
-	echo := examples.Echo{Namespace: "bookinfo"}
-	sleep := examples.Sleep{Namespace: "bookinfo"}
-	util.KubeDeleteContents("bookinfo", echoAllv1Yaml)
-	echo.Uninstall()
-	sleep.Uninstall()
-	time.Sleep(time.Duration(20) * time.Second)
-}
-
-func TestTCPShifting(t *testing.T) {
-	test.NewTest(t).Id("T4").Groups(test.Full, test.InterOp).NotRefactoredYet()
-
-	defer cleanupTCPShifting()
-	defer util.RecoverPanic(t)
-
-	log.Log.Info("TestTCPShifting")
-	echo := examples.Echo{Namespace: "bookinfo"}
-	sleep := examples.Sleep{Namespace: "bookinfo"}
-	echo.Install()
-	sleep.Install()
-
-	t.Run("TrafficManagement_100_percent_v1_tcp_shift_test", func(t *testing.T) {
-		defer util.RecoverPanic(t)
-		log.Log.Info("Shifting all TCP traffic to v1")
-		util.KubeApplyContents("bookinfo", echoAllv1Yaml)
-
-		sleepPod, err := util.GetPodName("bookinfo", "app=sleep")
-		util.Inspect(err, "Failed to get sleep pod name", "", t)
-		cmd := fmt.Sprintf(`sh -c "(date; sleep 1) | nc %s %s"`, "tcp-echo", "9000")
-		for i := 0; i < 20; i++ {
-			msg, err := util.PodExec("bookinfo", sleepPod, "sleep", cmd, true)
-			util.Inspect(err, "Failed to get response", "", t)
-			if !strings.Contains(msg, "one") {
-				t.Errorf("echo one; Got response: %s", msg)
-				log.Log.Errorf("echo one; Got response: %s", msg)
-			} else {
-				log.Log.Infof("%s", msg)
-			}
-		}
+// TestTcpTrafficShifting validates TCP traffic shifting feature.
+func TestTcpTrafficShifting(t *testing.T) {
+	test.NewTest(t).Id("T4").Groups(test.Full, test.InterOp).Run(func(t test.TestHelper) {
+		hack.DisableLogrusForThisTest(t)
+		ns := "foo"
+
+		t.Cleanup(func() {
+			oc.RecreateNamespace(t, ns)
```

Comment:
If you replace this with `app.Uninstall(t, app.Sleep(ns), app.EchoV1(ns), app.EchoV2(ns))`, the test takes less than 8 seconds!

But you have to fix tcp-echo-v1.yaml and tcp-echo-v2.yaml, because they both contain the same Service, so the Uninstall function fails because you're trying to delete the service, but the service doesn't exist.

Reply:
Nevermind about the tcp-echo-v1/v2.yaml files. Just approve #437 and the Uninstall function will work fine (it won't error even if you try to delete the same resource twice).

Reply:
If t.Fatalf occurred in one sub-test, will the execution keep running the sub-test t.Cleanup function ?

Reply:
Yes, the Cleanup function runs at the end of the subtest, regardless if it the test fails or not.